src/Extension/Engine/SiteTreePublishingEngine.php

```diff
@@ -36,130 +42,252 @@ class SiteTreePublishingEngine extends SiteTreeExtension implements Resettable
+
+        // Apparently ParentID can sometimes be string, so make sure we cast to (int) for our comparison
+        $parentChanged = $parentId && (int) $parentId !== (int) $owner->ParentID;
+        $urlSegmentChanged = $urlSegment && $original->URLSegment !== $owner->URLSegment;
+
         $context = [
-            'action' => 'publish',
+            'action' => self::ACTION_PUBLISH,
+            // If a URL change has been detected, then we need to force the recursive regeneration of all child
+            // pages
+            'urlSegmentChanged' => $parentChanged || $urlSegmentChanged,
         ];
+
+        // Collect any additional changes (noting that some could already have been added in onBeforePublishRecursive())
         $this->collectChanges($context);
+        // Flush any/all changes that we have detected
         $this->flushChanges();
     }
 
     public function onBeforeUnpublish()
     {
         $context = [
-            'action' => 'unpublish',
+            'action' => self::ACTION_UNPUBLISH,
         ];
+        // We'll collect these changes now, but they won't be actioned until onAfterUnpublish()
         $this->collectChanges($context);
     }
 
     public function onAfterUnpublish()
     {
+        // Flush any/all changes that we have detected
         $this->flushChanges();
     }
 
     /**
      * Collect all changes for the given context.
      *
      * @param array $context
+     * @return void
      */
     public function collectChanges($context)
     {
         Environment::increaseMemoryLimitTo();
         Environment::increaseTimeLimitTo();
 
-        if ($this->getOwner()->hasExtension(PublishableSiteTree::class)
-            || $this->getOwner() instanceof StaticPublishingTrigger
-        ) {
-            $toUpdate = $this->getOwner()->objectsToUpdate($context);
-            $this->setToUpdate($toUpdate);
+        Versioned::withVersionedMode(function () use ($context) {
+            $action = $context['action'];
+            // Collection of changes needs to happen within LIVE or DRAFT depending on the action context
 
-            $toDelete = $this->getOwner()->objectsToDelete($context);
-            $this->setToDelete($toDelete);
-        }
+            // Unpublish actions are called onBefore(), and their purpose is to remove the URLs of previously published
+            // pages. As such, we need to find out what the URL was/is in the current LIVE state (before the unpublish()
+            // completes)
+
+            // Publish actions are called onAfter(), and they need to retrieve whatever the current URL is in DRAFT.
+            // This is purely because if a page has an unpublished parent, then the LIVE URL will be incorrect (it will
+            // be missing the parent slug) - we'd prefer to cache to correct URL (with parentage) even though it'll be
+            // a cache of a 404
+            Versioned::set_stage($action === self::ACTION_UNPUBLISH ? Versioned::LIVE: Versioned::DRAFT);
```

Comment:
```suggestion
            $stage = $action === self::ACTION_UNPUBLISH ? Versioned::LIVE: Versioned::DRAFT;
            Versioned::set_reading_mode("Stage.$stage");
```
Since you're wrapping in Versioned::withVersionedMode() you should update the reading mode instead of just the stage

Reply:
@emteknetnz I'm looking at `Versioned::set_stage()`, and it looks like this also calls `set_reading_mode()`.

```php
public static function set_stage($stage)
{
    ReadingMode::validateStage($stage);
    static::set_reading_mode('Stage.' . $stage);
}
```

Whereas using `set_reading_mode()` **only** sets the reading mode, and doesn't also set the stage.